### PR TITLE
Deng convection scheme only works with MYJ or MYNN PBL, and deep convection must be turned off 

### PR DESCRIPTION
**Pull request by smileMchen**
TYPE:  Text only

KEYWORDS:  Deng shallow convection

SOURCE: Internal

DESCRIPTION OF CHANGES: 
The Deng shallow convection scheme only works with MYJ or MYNN PBL schemes. In addition, when this scheme is turned on, cumulus scheme must be turned off.
The README file which describes many of the namelist settings is updated to make 
this restriction clear. 

LIST OF MODIFIED FILES:
M   run/README.namelist
M  share/module_check_a_mundo.F

TESTS CONDUCTED: a test case with the wrong settings is conducted. Below is the error message
```
Quilting with   1 groups of   0 I/O tasks.
 Ntasks in X            2 , ntasks in Y            2
--- ERROR: Deng shallow convection can only work with MYJ or MYNN PBL
  --- ERROR: Fix shcu_physics or bl_pbl_physics in namelist.input.
--- ERROR: Deng shallow convection must run with cu_physics=0
  --- ERROR: Fix shcu_physics or cu_physics in namelist.input.
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE:  <stdin>  LINE:    1780
NOTE:       2 namelist settings are wrong. Please check and reset these options
-------------------------------------------
```

### CODE REVIEW

**Comment by davegill** [comment]
@smileMchen 
Is this restriction already in the check_a_mundo code?

With v4.1, we have quite a few schemes that need some documentation for the User's Guide.

**Comment by smileMchen** [comment]
@davegill 
No, it is not in check_a_mundo. I know this issue when I incorporate the Deng scheme to WRFV3.6 a few years ago.

**Comment by davegill** [comment]
@smileMchen 
Ming,
Since now this is no longer just a change to a README:
1. Change the title
2. Change the PR message (now there are two files, now there is a need for testing)
3. For a test, put in a section that shows what happens if you ask for Deng with an incorrect PBL option

File: run/README.namelist [hunk 1]
```diff
@@ -798,6 +798,7 @@ Namelist variables for controlling the adaptive time step option:
                                      = 2, Park and Bretherton shallow cumulus from CAM5 (CESM 1_0_1)
                                      = 3, GRIMS shallow cumulus from YSU group
                                      = 5, Deng shallow cumulus from PSU and WRF-Solar
+                                          This scheme only works with MYJ and MYNN PBL schemes
```
Comment (davegill) [inline run/README.namelist]: Deng shcu works only with MYJ or MYNN PBL schemes

File: share/module_check_a_mundo.F [hunk 2]
```diff
@@ -194,6 +194,20 @@ SUBROUTINE check_nml_consistency
...
          END IF
       ENDDO
+!-----------------------------------------------------------------------
+! Check that Deng Shallow Convection Must work with MYJ or MYNN PBL
+!-----------------------------------------------------------------------
+      DO i = 1, model_config_rec % max_dom
+         IF ( model_config_rec % shcu_physics(i) .EQ. 5 .AND. &
+              (model_config_rec % bl_pbl_physics(i) .NE. 2 .OR. &
+               model_config_rec % bl_pbl_physics(i) .NE. 5 ) ) THEN
+            wrf_err_message = '--- ERROR: Deng shallow convection can only work with MYJ and MYNN PBL '
```
Comment (davegill) [inline share/module_check_a_mundo.F]: Make this "and" an "or"

File: share/module_check_a_mundo.F [hunk 3]
```diff
@@ -194,6 +194,20 @@ SUBROUTINE check_nml_consistency
          count_fatal_error = count_fatal_error + 1
          END IF
       ENDDO
+!-----------------------------------------------------------------------
+! Check that Deng Shallow Convection Must work with MYJ or MYNN PBL
+!-----------------------------------------------------------------------
+      DO i = 1, model_config_rec % max_dom
+         IF ( model_config_rec % shcu_physics(i) .EQ. 5 .AND. &
```
Comment (davegill) [inline share/module_check_a_mundo.F]: Use the Registry generated names instead of 5, 2, 5

**Comment by davegill** [comment]
@smileMchen 
Ming,
We want to push this into release-v4.1, not develop.

**Comment by smileMchen** [comment]
@davegill @dudhia @weiwangncar 
Is it OK we change the Deng shallow convection scheme to convection scheme and rename the namelist option? It is quite confusing to name it ' shallow convection scheme' when it is actually a convection scheme that handles both shallow and deep convection.

File: run/README.namelist (hunk 1 above)
Comment (kkeene44) [reply run/README.namelist]: @davegill 
Is this comment resolved now?

**Comment by davegill** [comment]
@smileMchen @dudhia @weiwangncar 
Folks,
If this is a KF-variant deep-convection scheme which includes a shallow convection component, as Ming says, we could rename this. Make it a regular cumulus option. That means moving the code call, changing makefiles, changing the Registry, etc. This kind of restructuring actually seems like a good deal of work for us trying to release quickly.

I am not sure that there is a simple change can we make for v4.1 that will be the same user-option in v4.2.

We could announce in the v4.1 notes that for v4.2 and after, the Deng shallow scheme will be renamed and reclassified as the KFD cumulus scheme.

**Comment by smileMchen** [comment]
@davegill 
Dave,
I have modified README.namelist and module_check_a_mundo.F to make sure the Deng scheme only works with MYJ/MYNN PBL and cumulus scheme must be turned off. The commit message is updated to include error message with wrong PBL and cumulus settings when the Deng scheme is turned on.

**Comment by davegill** [comment]
@smileMchen 
Ming,
Please modify the PR commit message to include the changes to check_a_mundo.

**Comment by smileMchen** [comment]
@davegill 
Dave,
The commit message is updated.

File: share/module_check_a_mundo.F [hunk 4]
```diff
@@ -194,6 +194,28 @@ SUBROUTINE check_nml_consistency
...
+!-----------------------------------------------------------------------
+! Check that Deng Shallow Convection Must work with MYJ or MYNN PBL
+!-----------------------------------------------------------------------
+      DO i = 1, model_config_rec % max_dom
+         IF ( model_config_rec % shcu_physics(i) == dengshcuscheme .AND. &
+              (model_config_rec % bl_pbl_physics(i) /= myjpblscheme .OR. &
+               model_config_rec % bl_pbl_physics(i) /= mynnpblscheme2 ) ) THEN
+            wrf_err_message = '--- ERROR: Deng shallow convection can only work with MYJ or MYNN PBL '
+            CALL wrf_message ( wrf_err_message )
+            wrf_err_message = '--- ERROR: Fix shcu_physics and bl_pbl_physics in namelist.input.'
```
Comment (weiwangncar) [inline share/module_check_a_mundo.F]: Should be "Fix shcu_physics or bl_pbl_physics in namelist.input". They don't need to change both.

File: share/module_check_a_mundo.F [hunk 5]
```diff
@@ -194,6 +194,28 @@ SUBROUTINE check_nml_consistency
...
+         CALL wrf_debug ( 0, TRIM( wrf_err_message ) )
+         count_fatal_error = count_fatal_error + 1
+         END IF
+         IF ( model_config_rec % shcu_physics(i) == dengshcuscheme .AND. &
+              model_config_rec % cu_physics(i) /= nocuscheme  ) THEN
+            wrf_err_message = '--- ERROR: Deng shallow convection must run with cu_physics=0 '
+            CALL wrf_message ( wrf_err_message )
+            wrf_err_message = '--- ERROR: Fix shcu_physics and cu_physics in namelist.input.'
```
Comment (weiwangncar) [inline share/module_check_a_mundo.F]: @smileMchen The same here: "Fix shcu_physics or cu_physics in namelist.input". They don't need to change both.

**Pull request merged**